### ISSUE TITLE
[typescript-resolvers] Allow resolvers to return lazy parameters or promises

### ISSUE THREAD

**Issue by cranberyxl** [dotansimha/graphql-code-generator#3112]
**Is your feature request related to a problem? Please describe.**
I use a relay connection style in my schema. I have a single method that builds the connection fields, but returns them as functions like:

```
{
  pageInfo: () => PageInfo,
  count: () => number,
  edges: () => Edge[]
}
```
Unfortunately, the Resolver type expects them to be json or a promise returning json.

**Describe the solution you'd like**
I'd like to find a way that all of the fields can resolve as json, promise, or function without having a mapper for every connection. 

**Describe alternatives you've considered**
I looked at trying to return each of the connection fields as their own resolver, but I end up writing a lot more code and have to make a mapper for every connection. 

**Additional context**
Let me know if I'm overlooking a config, or way to writing this. 

Here's what I want my resolver to look like:
```
const eventsResolver: QueryResolvers['events'] = async (_, { input }) =>
  buildConnection(EventsQuery, input);
```

My Interfaces:
```
interface Node {
  id: ID!
}

interface Edge {
  cursor: String!
  node: Node!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
}

interface Connection {
  pageInfo: PageInfo!
  count: Int!
  edges: [Edge!]!
}
```

**Comment by dotansimha** [comment]
@cranberyxl You can use `defaultMapper` for that: https://graphql-code-generator.com/docs/plugins/typescript-resolvers#defaultmapper-string

**Comment by ardatan** [comment]
@cranberyxl You mean providing custom `return` and `parent` types or providing factory functions of that type.
For example you have `User` like below in your schema;
```gql
type User {
 id: ID
 name: String
 books: [Book]
}
```
and the resolvers are likebelow;
```ts
{
 Query: {
   user: (_, args) => {
      return {
        id: 1,
        name: 'Arda',
        books: () => getBooks() // So books are given as function to be called on demand.
      }
    }
 }
}
```

Is that what you asked or something else?

**Comment by cranberyxl** [comment]
@ardatan That's correct, because you can return any JSON from any resolver that satisfies it's children. Which may be a function or a promise.

**Comment by dotansimha** [comment]
@ardatan do you think it's something we can implement in the generated output?

**Comment by rintaun** [comment]
I think the underlying problem here is that, anywhere that can return a value could instead return:
* a Promise for that value
* a function that returns the value
* a function that returns a Promise for the value

All of which will function correctly. That's true of both resolvers and fields.

-----

I am actually running into a similar situation right now, where I have something like:

```graphql
type Story {
  id: ID!
  # ... irrelevant stuff ...
  isManageable: Boolean! # Whether it is manageable by the current user
}

type Query {
  story(id: ID!): Story
}
```

So when my resolver looks like this:

```typescript
// user model definition
const createUserModel = (context) => {
  return class UserModel {
    public static find(id: string): UserModel {}

    public id: string
    isManageable() { id == context.currentUserId }
  }
}

// context for reference
const context = {
  currentUserId,
  UserModel, // which has access to currentUserId
}

const Query: QueryResolvers = {
  story(_parent, { id }, { UserModel }, _info) {
    return UserModel.find(id)
    // return value is basically:
    { id: string, isManageable(): boolean }
  }
}
```

As far as I can tell, there isn't a type generated that can handle mixed values and functions like that, currently.

**Comment by rintaun** [comment]
Playing around with the generated types, I can accomplish what I believe to be the correct behavior, but it's a bit heavy-handed...

```typescript
export type FieldWrapper<T> = Promise<T> | T | (() => Promise<T> | T)

export interface Story {
  __typename?: 'Story'
  id: FieldWrapper<Scalars['ID']>
  isManageable: FieldWrapper<Scalars['Boolean']>
}
```

That would have to be done for every declared GraphQL type.

For the resolvers, interestingly `ResolverTypeWrapper` needs to be identical to `FieldWrapper`, above, and we need to tweak `Resolver` a bit:

```typescript
export type ResolverTypeWrapper<T> = Promise<T> | T | (() => Promise<T> | T)
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> =
  | ResolverFn<TResult, TParent, TContext, TArgs>
  | StitchingResolver<TResult, TParent, TContext, TArgs>
  | TResult
  | Promise<TResult>
```

**Comment by rintaun** [comment]
I started digging into fixing this, and the `ResolverTypeWrapper` fix can be applied with a config change:

```yaml
config:
  resolverTypeWrapperSignature: Promise<T> | T | (() => Promise<T> | T)
```

I'm working on a pull request (or two) for the `FieldWrapper` and `Resolver` fixes, which both seem to be relatively straightforward.

**Comment by rintaun** [comment]
@cranberyxl I believe that #3299 would allow you to solve your problem. what do you think?

**Comment by cranberyxl** [comment]
@rintaun Looks good. I'll be excited to try it. We've been adding many mappers which this should help us avoid.

**Comment by dotansimha** [comment]
Fixed in `v1.13.0`, thanks @rintaun
